substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/code/CodeInfoEncoder.java

```diff
@@ -393,6 +577,286 @@ public boolean verifyFrameInfo(CodeInfo info) {
+     * instead of through their type signature. Primitive values are written directly into the
+     * encoding. This means that our encoding can be of a different length from the JDK one.
+     * 
+     * We use a modified version of the ConstantPool and AnnotationParser classes to decode the
+     * data, since those are not used in their original functions at runtime. (see
+     * {@link com.oracle.svm.core.jdk.Target_jdk_internal_reflect_ConstantPool})
+     */
+    class AnnotationEncoder {
+        byte[] encodeAnnotations(Annotation[] annotations) throws InvocationTargetException, IllegalAccessException {
+            UnsafeArrayTypeWriter buf = UnsafeArrayTypeWriter.create(ByteArrayReader.supportsUnalignedMemoryAccess());
+
+            Annotation[] filteredAnnotations = filterAnnotations(annotations);
+            buf.putU2(filteredAnnotations.length);
+            for (Annotation annotation : filteredAnnotations) {
+                encodeAnnotation(buf, annotation);
+            }
+
+            return buf.toArray();
+        }
+
+        byte[] encodeParameterAnnotations(Annotation[][] annotations) throws InvocationTargetException, IllegalAccessException {
+            UnsafeArrayTypeWriter buf = UnsafeArrayTypeWriter.create(ByteArrayReader.supportsUnalignedMemoryAccess());
+
+            buf.putU1(annotations.length);
+            for (Annotation[] parameterAnnotations : annotations) {
+                Annotation[] filteredParameterAnnotations = filterAnnotations(parameterAnnotations);
+                buf.putU2(filteredParameterAnnotations.length);
+                for (Annotation parameterAnnotation : filteredParameterAnnotations) {
+                    encodeAnnotation(buf, parameterAnnotation);
+                }
+            }
+
+            return buf.toArray();
+        }
+
+        void encodeAnnotation(UnsafeArrayTypeWriter buf, Annotation annotation) throws InvocationTargetException, IllegalAccessException {
+            buf.putU2(encoders.sourceClasses.getIndex(annotation.annotationType()));
+            AnnotationType type = AnnotationType.getInstance(annotation.annotationType());
+            buf.putU2(type.members().size());
+            for (Map.Entry<String, Method> entry : type.members().entrySet()) {
+                String memberName = entry.getKey();
+                Method valueAccessor = entry.getValue();
+                buf.putU2(encoders.sourceMethodNames.getIndex(memberName));
+                encodeValue(buf, valueAccessor.invoke(annotation), type.memberTypes().get(memberName));
+            }
+        }
+
+        void encodeValue(UnsafeArrayTypeWriter buf, Object value, Class<?> type) throws InvocationTargetException, IllegalAccessException {
+            buf.putU1(tag(type));
+            if (type.isAnnotation()) {
+                encodeAnnotation(buf, (Annotation) value);
+            } else if (type.isEnum()) {
+                buf.putU2(encoders.sourceClasses.getIndex(type));
```

Comment:
We can have more than 65000 classes in an image. And also more than 65000 entries in `sourceMethodNames` for the `String` case below. So all of these cannot use `U2` encoding.

Reply:
I ran into that problem on JCK gates, I'm in the process of fixing it.

Reply:
Fixed